### DISCUSSION TITLE
How to calculate Green function in 2D Fermi-Hubbard model ?

### DISCUSSION THREAD

**Discussion by xuejiayu** [QuSpin/QuSpin#440] [Q&A]
Dear QuSpin team,
         Thanks for your wonderful software about ED ! I have a question about it:
I want to calculate Green function <c^{\dagger}_{i,up}c_{j,up}> in the half filled 2D spinful Fermi Hubbard model 
my basis setting is : basis_2d=spinful_fermion_basis_general(N_2d,Nf=(int(N_2d/2),int(N_2d/2))) (in Nf=(N_2d/2,N_2d/2) sector, we can find ground state). After construct the model, I can calculate green function(<c_{r,up}^{\dagger}c_{r',up}>) as following way : 
 Gr_static=[[1.0,r,rp]] # <c_{r,up}^{\dagger}c_{r',up}>
 Gr_op=hamiltonian([["+-|",Gr_static]],[],basis=basis_2d, check_herm=False,check_symm=False,check_pcon=False)
 Gr=Gr_op.matrix_ele(Vl, Vr,time=0,diagonal=False) (Vl,Vr is lef and right eigenvector, respectively)
my result is shown that  G(i,j)=0 if i/=j. This result is unphysical. So I want to ask what is the correct way to compute Green function in Hubbard model ? Many thanks !  
Best regards,
Sugar

**Comment by JiahaoYao** [comment]
@xuejiayu can you post the code you use to compute for the experiments so we can use this as a reference in debugging. For example, how do u define `r` and `rp`?

**Reply by xuejiayu** [reply]
Dear jiahao, thanks for the comment. here is my naive code: 

###### define model parameters ######
Lx, Ly = 2,2 # linear dimension of 2d honeycomb lattice
LQ=Lx*Ly # number of unit cell 
Norb=2 # 0: A sublattice 1: B sublattice 
N_2d =Norb*Lx*Ly # number of sites for honeycomb lattice
# Hubbard model: H=-t\sum_{<ij>}(c^{\dagger}_{i}c_{j}+h.c)+U\sum_{i}(n_{i,up}-1/2)(n_{i,dn}-1/2)
t=1.0 # hopping matrix element
delta=0.0 # hopping correction
U=4.0 # onsite interaction
mu=U*0.5
#
# define the honeycomb lattice 
LIST=np.zeros((LQ,2))
NLIST=np.zeros((N_2d,3))
INVNLIST=np.zeros((Lx,Ly,Norb))
Invlist=np.zeros((LQ,LQ))
L_bonds=np.zeros((LQ,4))


nc=0
for NX in range(Lx):
    for NY in range(Ly):
        LIST[nc][0]=NX
        LIST[nc][1]=NY
        Invlist[NX][NY]=nc
        nc=nc+1
nc=0
for nx in range(Lx):
    for ny in range(Ly):
        for NO in range(Norb):   # notice NO=0 for A sub lattice, NO=1 for B sub lattice
            NLIST[nc][0]=nx
            NLIST[nc][1]=ny
            NLIST[nc][2]=NO
            INVNLIST[nx][ny][NO]=nc
            nc=nc+1

for iy in range(Ly):
    for ix in range(Lx):
            ll=int(Invlist[ix][iy])
            L_bonds[ll][0]=INVNLIST[ix][iy][0]
            L_bonds[ll][1]=INVNLIST[ix][iy][1]
            lln1=INVNLIST[ix][(iy-1)%Ly][1]
            L_bonds[ll][2]=lln1
            lln2=INVNLIST[(ix+1)%Lx][(iy-1)%Ly][1]
            L_bonds[ll][3]=lln2

## construct the model  basis
basis_2d=spinful_fermion_basis_general(N_2d,Nf=(int(N_2d/2),int(N_2d/2)))
#
###### setting up hamiltonian ######
# setting up site-coupling lists for simple case
hopping_ij1up = [[-(t+delta),L_bonds[i][0],L_bonds[i][1]] for i in range(LQ)]
hopping_ij2up = [[-(t+delta),L_bonds[i][0],L_bonds[i][2]] for i in range(LQ)]
hopping_ij3up = [[-(t+delta),L_bonds[i][0],L_bonds[i][3]] for i in range(LQ)]

hopping_ji1up = [[+(t-delta),L_bonds[i][0],L_bonds[i][1]] for i in range(LQ)]
hopping_ji2up = [[+(t-delta),L_bonds[i][0],L_bonds[i][2]] for i in range(LQ)]
hopping_ji3up = [[+(t-delta),L_bonds[i][0],L_bonds[i][3]] for i in range(LQ)]

hopping_ij1dn = [[-(t-delta),L_bonds[i][0],L_bonds[i][1]] for i in range(LQ)]
hopping_ij2dn = [[-(t-delta),L_bonds[i][0],L_bonds[i][2]] for i in range(LQ)]
hopping_ij3dn = [[-(t-delta),L_bonds[i][0],L_bonds[i][3]] for i in range(LQ)]

hopping_ji1dn = [[+(t+delta),L_bonds[i][0],L_bonds[i][1]] for i in range(LQ)]
hopping_ji2dn = [[+(t+delta),L_bonds[i][0],L_bonds[i][2]] for i in range(LQ)]
hopping_ji3dn = [[+(t+delta),L_bonds[i][0],L_bonds[i][3]] for i in range(LQ)]

interaction=[[U,i,i] for i in range(N_2d)]
chemical_potential=[[-mu,i] for i in range(N_2d)]
#
print(t-delta,t+delta)
static=[["+-|",hopping_ij1up], # spin up hops to left
        ["+-|",hopping_ij2up],
        ["+-|",hopping_ij3up],
        ["-+|",hopping_ji1up],
        ["-+|",hopping_ji2up],
        ["-+|",hopping_ji3up],
		["|+-",hopping_ij1dn], # spin down hopes to left
        ["|+-",hopping_ij2dn],
        ["|+-",hopping_ij3dn],
		["|-+",hopping_ji1dn], # spin up hops to right
        ["|-+",hopping_ji2dn],
        ["|-+",hopping_ji3dn],
        ["n|",chemical_potential],
        ["|n",chemical_potential],
		["n|n",interaction]] # spin up-spin down interaction
# build hamiltonian
no_checks=dict(check_pcon=False,check_herm=False,check_symm=False)
H=hamiltonian(static,[],basis=basis_2d,dtype=np.complex128,**no_checks)
# diagonalise H
Ham=H.toarray()
#E=H.eigvalsh()
E,vecl,vecr=sc.linalg.eig(Ham,left=True, right=True)
Vl=vecl[:,0]
Vr=vecr[:,0]

E=E+U*N_2d*1.0/4

Gr=np.zeros((N_2d,N_2d),dtype=complex)

#----Green function--- only for A sublattice # 
for ix in range(Lx):
    for iy in range(Ly):
        r=int(INVNLIST[ix][iy][0])
        for jx in range(Lx):
            for jy in range(Ly):
                rp=int(INVNLIST[jx][jy][0]) # r,rp are two different sites in the lattice 
                Gr_static=[[1.0,r,rp]] # <c_{r,up}^{\dagger}c_{r',up}>
                Gr_op=hamiltonian([["+-|",Gr_static]],[],basis=basis_2d, check_herm=False,check_symm=False,check_pcon=False)
                Gr[r,rp]=Gr_op.matrix_ele(Vl, Vr,time=0,diagonal=False)

Let me know if you need more explanation for the code

**Reply by JiahaoYao** [reply]
thanks @xuejiayu , received. will take a look, and might get back to you at a later time this week.

**Reply by JiahaoYao** [reply]
Hi @xuejiayu, your code looks quite nice. 

First of all, thanks for choosing to use `Quspin`!

Here are several questions: 

1. i assume you are working on the periodic boundary case. 

Sometimes, there is an issue for the `L=2` case. For example, when you mentioned `iy` and `(iy-1)%Ly` and loop `iy in range(2)`, then the connections will be `0 --- 1` and `1 --- 0` (this one might be artificial since `1` gets back to `0`).

if this is the case, you might need to change 
```python
hopping_ij1up = [[-(t+delta),L_bonds[i][0],L_bonds[i][1]] for i in range(LQ)]
hopping_ij2up = [[-(t+delta),L_bonds[i][0],L_bonds[i][2]] for i in range(LQ)]
hopping_ij3up = [[-(t+delta),L_bonds[i][0],L_bonds[i][3]] for i in range(LQ)]

hopping_ji1up = [[+(t-delta),L_bonds[i][0],L_bonds[i][1]] for i in range(LQ)]
hopping_ji2up = [[+(t-delta),L_bonds[i][0],L_bonds[i][2]] for i in range(LQ)]
hopping_ji3up = [[+(t-delta),L_bonds[i][0],L_bonds[i][3]] for i in range(LQ)]

hopping_ij1dn = [[-(t-delta),L_bonds[i][0],L_bonds[i][1]] for i in range(LQ)]
hopping_ij2dn = [[-(t-delta),L_bonds[i][0],L_bonds[i][2]] for i in range(LQ)]
hopping_ij3dn = [[-(t-delta),L_bonds[i][0],L_bonds[i][3]] for i in range(LQ)]

hopping_ji1dn = [[+(t+delta),L_bonds[i][0],L_bonds[i][1]] for i in range(LQ)]
hopping_ji2dn = [[+(t+delta),L_bonds[i][0],L_bonds[i][2]] for i in range(LQ)]
hopping_ji3dn = [[+(t+delta),L_bonds[i][0],L_bonds[i][3]] for i in range(LQ)]
```

to 
```python 
hopping_ij1up = [[-(t+delta),L_bonds[i][0],L_bonds[i][1]] for i in range(LQ)]
hopping_ij2up = [[-(t+delta),L_bonds[i][0],L_bonds[i][2]] for i in range(LQ) if NLIST[i][1] == 1]
hopping_ij3up = [[-(t+delta),L_bonds[i][0],L_bonds[i][3]] for i in range(LQ) if NLIST[i][1] == 1 and NLIST[i][0] == 0 ]

hopping_ji1up = [[+(t-delta),L_bonds[i][0],L_bonds[i][1]] for i in range(LQ)]
hopping_ji2up = [[+(t-delta),L_bonds[i][0],L_bonds[i][2]] for i in range(LQ) if NLIST[i][1] == 1]
hopping_ji3up = [[+(t-delta),L_bonds[i][0],L_bonds[i][3]] for i in range(LQ) if NLIST[i][1] == 1 and NLIST[i][0] == 0 ]

hopping_ij1dn = [[-(t-delta),L_bonds[i][0],L_bonds[i][1]] for i in range(LQ)]
hopping_ij2dn = [[-(t-delta),L_bonds[i][0],L_bonds[i][2]] for i in range(LQ) if NLIST[i][1] == 1]
hopping_ij3dn = [[-(t-delta),L_bonds[i][0],L_bonds[i][3]] for i in range(LQ) if NLIST[i][1] == 1 and NLIST[i][0] == 0 ]

hopping_ji1dn = [[+(t+delta),L_bonds[i][0],L_bonds[i][1]] for i in range(LQ)]
hopping_ji2dn = [[+(t+delta),L_bonds[i][0],L_bonds[i][2]] for i in range(LQ) if NLIST[i][1] == 1]
hopping_ji3dn = [[+(t+delta),L_bonds[i][0],L_bonds[i][3]] for i in range(LQ) if NLIST[i][1] == 1 and NLIST[i][0] == 0 ]
```

(but it seems like not this issue, so please just ignore this.)

2. check your symmetry with the model in mind, i.e. `spinful_fermion_basis_general(N_2d,Nf=(int(N_2d/2),int(N_2d/2)))`  

3.
```python
E,vecl,vecr=sc.linalg.eig(Ham,left=True, right=True)
```
This is quite time-consuming. 

recommend to change to 
```python
Er, Vr = H.eigsh(k=4,which='SA',maxiter=1E4)
El, Vl = H.T.eigsh(k=4,which='SA',maxiter=1E4)
Vr=Vr[:, 0]
Vl=Vl[:, 0]
```

------
You might also find [this](http://weinbe58.github.io/QuSpin/examples/example18.html) interesting, where the authors of quspin already provides a nice way to figure out the connections of the honeycomb; probably you might not need to figure out the index by hand.

**Reply by JiahaoYao** [reply]
by the way, may i ask for the definition of the Green's function for the hubbard model, in the code you mentioned `<c_{r,up}^{\dagger}c_{r',up}>`, what's for the expectation is with respect to?

More precisely, I am not sure which one is correctly? 
```python
Gr_op.matrix_ele(Vl, Vr,time=0,diagonal=False)
```

or 
```python
Gr_op.matrix_ele(Vr, Vr,time=0,diagonal=False)
```

**Reply by JiahaoYao** [reply]
As for the item `2`, i change the `Nf` to `basis_2d=spinful_fermion_basis_general(N_2d,Nf=(2, 2))`

Run ur your code, you can find Gr by adding this at the end

```python
                if r != rp: 
                    print(Gr[r,rp])
```


i.e. 

```python

import numpy as np 
import scipy as sc 
from quspin.basis import spinful_fermion_basis_general
from quspin.operators import hamiltonian

###### define model parameters ######
Lx, Ly = 2,2 # linear dimension of 2d honeycomb lattice
LQ=Lx*Ly # number of unit cell 
Norb=2 # 0: A sublattice 1: B sublattice 
N_2d =Norb*Lx*Ly # number of sites for honeycomb lattice
# Hubbard model: H=-t\sum_{<ij>}(c^{\dagger}_{i}c_{j}+h.c)+U\sum_{i}(n_{i,up}-1/2)(n_{i,dn}-1/2)
t=1.0 # hopping matrix element
delta=0.0 # hopping correction
U=4.0 # onsite interaction
mu=U*0.5
#
# define the honeycomb lattice 
LIST=np.zeros((LQ,2))
NLIST=np.zeros((N_2d,3))
INVNLIST=np.zeros((Lx,Ly,Norb))
Invlist=np.zeros((LQ,LQ))
L_bonds=np.zeros((LQ,4))


nc=0
for NX in range(Lx):
    for NY in range(Ly):
        LIST[nc][0]=NX
        LIST[nc][1]=NY
        Invlist[NX][NY]=nc
        nc=nc+1
nc=0
for nx in range(Lx):
    for ny in range(Ly):
        for NO in range(Norb):   # notice NO=0 for A sub lattice, NO=1 for B sub lattice
            NLIST[nc][0]=nx
            NLIST[nc][1]=ny
            NLIST[nc][2]=NO
            INVNLIST[nx][ny][NO]=nc
            nc=nc+1

for iy in range(Ly):
    for ix in range(Lx):
            ll=int(Invlist[ix][iy])
            L_bonds[ll][0]=INVNLIST[ix][iy][0]
            L_bonds[ll][1]=INVNLIST[ix][iy][1]
            lln1=INVNLIST[ix][(iy-1)%Ly][1]
            L_bonds[ll][2]=lln1
            lln2=INVNLIST[(ix+1)%Lx][(iy-1)%Ly][1]
            L_bonds[ll][3]=lln2

## construct the model  basis
basis_2d=spinful_fermion_basis_general(N_2d,Nf=(2, 2))
#
###### setting up hamiltonian ######
# setting up site-coupling lists for simple case
hopping_ij1up = [[-(t+delta),L_bonds[i][0],L_bonds[i][1]] for i in range(LQ)]
hopping_ij2up = [[-(t+delta),L_bonds[i][0],L_bonds[i][2]] for i in range(LQ)]
hopping_ij3up = [[-(t+delta),L_bonds[i][0],L_bonds[i][3]] for i in range(LQ)]

hopping_ji1up = [[+(t-delta),L_bonds[i][0],L_bonds[i][1]] for i in range(LQ)]
hopping_ji2up = [[+(t-delta),L_bonds[i][0],L_bonds[i][2]] for i in range(LQ)]
hopping_ji3up = [[+(t-delta),L_bonds[i][0],L_bonds[i][3]] for i in range(LQ)]

hopping_ij1dn = [[-(t-delta),L_bonds[i][0],L_bonds[i][1]] for i in range(LQ)]
hopping_ij2dn = [[-(t-delta),L_bonds[i][0],L_bonds[i][2]] for i in range(LQ)]
hopping_ij3dn = [[-(t-delta),L_bonds[i][0],L_bonds[i][3]] for i in range(LQ)]

hopping_ji1dn = [[+(t+delta),L_bonds[i][0],L_bonds[i][1]] for i in range(LQ)]
hopping_ji2dn = [[+(t+delta),L_bonds[i][0],L_bonds[i][2]] for i in range(LQ)]
hopping_ji3dn = [[+(t+delta),L_bonds[i][0],L_bonds[i][3]] for i in range(LQ)]



interaction=[[U,i,i] for i in range(N_2d)]
chemical_potential=[[-mu,i] for i in range(N_2d)]
#
print(t-delta,t+delta)
static=[["+-|",hopping_ij1up], # spin up hops to left
        ["+-|",hopping_ij2up],
        ["+-|",hopping_ij3up],
        ["-+|",hopping_ji1up],
        ["-+|",hopping_ji2up],
        ["-+|",hopping_ji3up],
		["|+-",hopping_ij1dn], # spin down hopes to left
        ["|+-",hopping_ij2dn],
        ["|+-",hopping_ij3dn],
		["|-+",hopping_ji1dn], # spin up hops to right
        ["|-+",hopping_ji2dn],
        ["|-+",hopping_ji3dn],
        ["n|",chemical_potential],
        ["|n",chemical_potential],
		["n|n",interaction]] # spin up-spin down interaction
# build hamiltonian
# no_checks=dict(check_pcon=False,check_herm=False,check_symm=False)
H=hamiltonian(static,[],basis=basis_2d,dtype=np.complex128)
# diagonalise H
Ham=H.toarray()
print(Ham)
#E=H.eigvalsh()

Er, Vr = H.eigsh(k=4,which='SA',maxiter=1E4)
El, Vl = H.T.eigsh(k=4,which='SA',maxiter=1E4)
Vr=Vr[:, 0]
Vl=Vl[:, 0]


# E=E+U*N_2d*1.0/4

Gr=np.zeros((N_2d,N_2d),dtype=complex)

#----Green function--- only for A sublattice # 
for ix in range(Lx):
    for iy in range(Ly):
        r=int(INVNLIST[ix][iy][0])
        for jx in range(Lx):
            for jy in range(Ly):
                rp=int(INVNLIST[jx][jy][0]) # r,rp are two different sites in the lattice 
                Gr_static=[[1.0,r,rp]] # <c_{r,up}^{\dagger}c_{r',up}>
                Gr_op=hamiltonian([["+-|",Gr_static]],[],basis=basis_2d, check_herm=False,check_symm=False,check_pcon=False)

                Gr[r,rp]=Gr_op.matrix_ele(Vl, Vr,time=0,diagonal=False)

                if r != rp: 
                    print(Gr[r,rp])
```

You will get 
```python 
(0.08454090047497861-1.366962099069724e-14j)
(0.10954408571479902-2.7332303087490573e-14j)
(0.0333452098747638+4.313216450668733e-14j)
(0.08454090047497861+1.366962099069724e-14j)
(0.033345209874771714+3.4741307053387516e-14j)
(0.10954408571476157-5.434541705540141e-14j)
(0.10954408571479901+2.7332303087490573e-14j)
(0.033345209874771714-3.473263343600763e-14j)
(0.08454090047491168+8.881784197001252e-16j)
(0.03334520987476379-4.3135633953639285e-14j)
(0.10954408571476157+5.434541705540141e-14j)
(0.08454090047491168-8.881784197001252e-16j)
```

**Reply by JiahaoYao** [reply]
if you are using the example in the tutorial, you might be interested in trying this 

```python
from quspin.basis import spinful_fermion_basis_general
from quspin.operators import hamiltonian
import numpy as np
import networkx as nx # networkx package, see https://networkx.github.io/documentation/stable/
import matplotlib.pyplot as plt # plotting library
#
###### create honeycomb lattice
# lattice graph parameters
m = 2  # number of rows of hexagons in the lattice
n = 2  # number of columns of hexagons in the lattice
isPBC = False # if True, use periodic boundary conditions
#
### build graph using networkx
hex_graph = nx.generators.lattice.hexagonal_lattice_graph(m, n, periodic=isPBC)
# label graph nodes by consecutive integers
hex_graph = nx.convert_node_labels_to_integers(hex_graph)
# set number of lattice sites
N = hex_graph.number_of_nodes()
print('constructed hexagonal lattice with {0:d} sites.\n'.format(N))
# visualise graph
pos = nx.spring_layout(hex_graph, seed=42, iterations=100)
nx.draw(hex_graph, pos=pos, with_labels=True)
plt.show(block=False)
#
###### model parameters
#
N_up = 2 # number of spin-up fermions
N_down = 2 # number of spin-down fermions
t = 1.0 # tunnelling matrix element
U = 2.0 # on-site fermion interaction strength
#
##### set up Fermi-Hubbard Hubbard Hamiltonian with quspin #####
#
### compute basis
basis = spinful_fermion_basis_general(N, Nf=(N_up, N_down))
print('Hilbert space size: {0:d}.\n'.format(basis.Ns))
#
# define site-coupling lists
tunnelling   = [[-t, i, j] for i in range(N) for j in hex_graph.adj[i]]
interactions = [[ U, i, i] for i in range(N)]
#
# define site-coupling lists [hermitian conjugates "-+|" and "|-+" contained in tunnelling list]
static = [["n|n", interactions], ["+-|", tunnelling], ["|+-", tunnelling]]
dynamic=[]
#
### construct Hamiltonian
H = hamiltonian(static, dynamic, basis=basis, dtype=np.float64)
#
# compute eigensystem
E, V = H.eigsh(k=4,which='SA',maxiter=1E4)
print(f'\nlowest energies: {E}')

sublattice_A = [i for i in range(N) if i % 2 == 0]
sublattice_B = [i for i in range(N) if i % 2 == 1]

Er, Vr = H.eigsh(k=4,which='SA',maxiter=1E4)
El, Vl = H.T.eigsh(k=4,which='SA',maxiter=1E4)
Vr=Vr[:, 0]
Vl=Vl[:, 0]

Gr=np.zeros((N,N),dtype=complex)

#----Green function--- only for A sublattice # 

for r in sublattice_A: 
    for rp in sublattice_A: 
        Gr_static=[[1.0,r,rp]] # <c_{r,up}^{\dagger}c_{r',up}>
        Gr_op=hamiltonian([["+-|",Gr_static]],[],basis=basis, check_herm=False,check_symm=False,check_pcon=False)

        Gr[r,rp]=Gr_op.matrix_ele(Vl, Vr,time=0,diagonal=False)

        if r != rp: 
            # print(Gr_op.dot(Vr)) 
            print(Gr[r,rp])
```

You can get 

```markdown
(-0.12666693031015683+0j)
(-0.0011919855926290424+0j)
(-0.12666693031015777+0j)
(0.002087245634555113+0j)
(0.061765782061435334+0j)
(-0.0011919855926300319+0j)
(0.061765782061434744+0j)
(-0.12666693031015944+0j)
(-0.03780263045032276+0j)
(-0.1618469521664953+0j)
(-0.05797626984381228+0j)
(0.029390459331741496+0j)
(-0.03316152675911643+0j)
(0.031577838995079884+0j)
(-0.0011919855926331591+0j)
(-0.03780263045032749+0j)
(-0.033161526759120735+0j)
(-0.0971347519298403+0j)
(-0.06929882173418367+0j)
(-0.05295939773545341+0j)
(-0.0645844977613653+0j)
(-0.1266669303101549+0j)
(-0.16184695216649006+0j)
(-0.033161526759117294+0j)
(-0.05797626984381557+0j)
(0.03157783899507537+0j)
(-0.037802630450325586+0j)
(0.029390459331736066+0j)
(0.002087245634555276+0j)
(-0.05797626984381354+0j)
(-0.09713475192984239+0j)
(-0.057976269843814224+0j)
(-0.11875939743397118+0j)
(-0.09713475192984247+0j)
(-0.11875939743397085+0j)
(0.06176578206143406+0j)
(0.029390459331737617+0j)
(-0.06929882173418736+0j)
(0.03157783899507653+0j)
(-0.1187593974339707+0j)
(-0.06458449776136849+0j)
(-0.116690374164855+0j)
(-0.0011919855926260312+0j)
(-0.03316152675911349+0j)
(-0.05295939773545562+0j)
(-0.037802630450321034+0j)
(-0.09713475192984453+0j)
(-0.06458449776137239+0j)
(-0.06929882173419058+0j)
(0.06176578206143672+0j)
(0.031577838995079155+0j)
(-0.06458449776136972+0j)
(0.029390459331740264+0j)
(-0.1187593974339721+0j)
(-0.11669037416485809+0j)
(-0.0692988217341877+0j)
```

**Reply by xuejiayu** [reply]
Hi Jihao,
 Thanks for your wonderful reply. Here has some problems I find :
1) Yes ! my Green function definition is : Gr(i,j)=<Vl|c^{\dagger}_{I,up}c_{up,dn}|Vr> (in our hermitian case, Vl=Vr)
2) According to your suggestion, I get different Vr,Vl  when I run the code every time. It seems like not converge in the calculation? 
3) I only care **half filled case**, so my basis setting is just : basis_2d=spinful_fermion_basis_general(N_2d,Nf=(4,4)) (in our 2*2 honeycomb lattice, i.e. 8 sites. I can use the example but I think my honeycomb lattice construction is OK). In this case, I get Gr(I,j)=0 if I/=j, but if Nf=(2,2), Gr(I,j) is finite for I/=j. 
Sugar

**Reply by JiahaoYao** [reply]
Hi @xuejiayu ，
1. i see, when you do the full diagonalization, do u find they are the same? then that is fine.
2. okay, then does that mean one has to full diagonalize the hamiltonian though costly. 
3. Do u mean `Gr(i,j)` should not be zero when you use `basis_2d=spinful_fermion_basis_general(N_2d,Nf=(4,4))`? Then, question is what might be wrong?

**Reply by xuejiayu** [reply]
Hi Jiahao,
1) The full diagonalization results are the same .
3) Physically, the Gr(I,j) should be finite for I/=j, so I think my results is unphysical.

**Reply by JiahaoYao** [reply]
Hi @xuejiayu, i see. Thanks.

**Comment by JiahaoYao** [comment]
I failed to figure out why `G(i, j) == 0` when `i≠j`. Is there someone else who happens to know this?
